[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test

Amyh11325 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61594**

### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

Commit:
[3/N] Nnapi Delegate: OSS Tests

Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

TODO:
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

CMakeLists.txt:
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's
use of pybind

test_backends.py:
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the
need to change dtype from float64 to float32

Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

[ghstack-poisoned]

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

[ghstack-poisoned]

Amyh11325 commented:
@Amyh11325 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29687143).

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

[ghstack-poisoned]

Amyh11325 commented:
@Amyh11325 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29687143).

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

[ghstack-poisoned]

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

[ghstack-poisoned]

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

[ghstack-poisoned]

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

[ghstack-poisoned]

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

[ghstack-poisoned]

Amyh11325 commented:
@Amyh11325 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29687143).

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

[ghstack-poisoned]

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

[ghstack-poisoned]

Amyh11325 commented:
@Amyh11325 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29687143).

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

[ghstack-poisoned]

Amyh11325 commented:
@Amyh11325 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29687143).

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

[ghstack-poisoned]

Amyh11325 commented:
@Amyh11325 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29687143).

raziel commented on test/jit/test_backends.py:
```diff
@@ -3,6 +3,7 @@
 import os
 import sys
 import unittest
+import ctypes
```
nit: sort alphabetically

raziel commented on test/jit/test_backends.py:
```diff
@@ -685,3 +687,75 @@ def test_execution(self):
     @skipIfRocm
     def test_errors(self):
         self.error_module_compiler_test.test_errors()
+
```
Should we move these tests to a separate file?

Thus far the tests in text_backends.py are not backend specific, and focus on the mechanism rather than any specific backend implementation.
Thus, I would prefer having either a single test file per backend (test_nnapi_backend.py), or if you think many backends can share a common test class then have one file (unfortunately the better name test_backends.py is used, so maybe test_backend_impls.py).

What do you think @iseeyuan ?

iseeyuan replied:
Makes sense to me. Probably it could also resolve the issue of int 32 type issue.

Amyh11325 replied:
It did not resolve the issue, sadly.

raziel commented on test/jit/test_backends.py:
```diff
@@ -685,3 +687,75 @@ def test_execution(self):
+        # Avoid saturation in fbgemm
+        torch.backends.quantized.engine = 'qnnpack'
+
+        libneuralnetworks_path = os.environ.get("LIBNEURALNETWORKS_PATH")
```
Just curious: why is this needed?

And should it fail if it can't?

Amyh11325 replied:
Perhaps, @axitkhurana or @dreiss can provide more insight. I've tried only converting to the NNAPI backend without these lines. It seemed run successfully, but I believe they are more necessary for execution, especially the loading of libneuralnetworks, if you're also referring to that.

axitkhurana replied:
Yes, we use it to test nnapi execution on CPU and compare the results with pytorch, conversion should work without it

raziel commented on test/jit/test_backends.py:
```diff
@@ -685,3 +687,75 @@ def test_execution(self):
+        # Save default dtype
+        module = torch.nn.PReLU()
+        default_dtype = module.weight.dtype
+        # Change dtype to float32 (since float 64 is not supported)
```
This is odd, can you provide more context on why is this needed (which still sounds worrisome as a workaround).

Amyh11325 replied:
It seems that the default dtype has been changed to float64 for this file. The problem seems specific to test_backends.py, since when I wrote a local script, I didn't need to do this. 
Since dtype is global, I think some other test is setting the dtype to float64. For example Akshit found this one, https://www.internalfb.com/code/fbsource/[a46774f32bd8]/fbcode/caffe2/test/jit/test_complexity.py?lines=8.
I was also running into problems with other tests after setting dtype to float32, so I had to add code to change it back to float64.

axitkhurana replied:
@Amyh11325 mind setting it back after the test in test_complexity and seeing if we still need it?

Amyh11325 replied:
I can try, but I think it would break other tests that require the default dtype to be float64, like when I didn't change dtype back to float64 in this PR.

raziel commented on test/jit/test_backends.py:
```diff
@@ -685,3 +687,75 @@ def test_execution(self):
+# First skip is needed for IS_WINDOWS or IS_MACOS to skip the tests.
+# Second skip is because ASAN is currently causing an error.
+# It is still unclear how to resolve this.
+# See https://fb.workplace.com/groups/1144215345733672/permalink/2010505639104634/
```
Maybe replace this with a task?
They're still opaque in OSS code but I think better than links to internal forums.

raziel reviewed (commented):
Looks great. I have some comments worth addressing though.

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

[ghstack-poisoned]

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

[ghstack-poisoned]

iseeyuan reviewed (changes requested):
LGTM, except for putting the tests in another file, like test_nnapi_delegate.py or something like that.

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

[ghstack-poisoned]

Amyh11325 commented:
@Amyh11325 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29687143).

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

[ghstack-poisoned]

Amyh11325 commented:
@Amyh11325 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29687143).

iseeyuan reviewed (approved):
LGTM. Thanks for the updates!

Commit:
Update on "[3/N] Nnapi Backend Delegate Preprocess: Basic OSS Test"


### Summary:
Added a unit test for the Nnapi delegate's preprocess() function. The
function was previously tested locally, but now a basic test is
added for OSS.

See https://github.com/pytorch/pytorch/pull/61499 for preprocess
implementation. See D29647123 for local testing.

**TODO:**
Add more comprehensive tests.
Add tests for model execution, after the Nnapi delegate's initialization
and execution is implemented T91991928.

**CMakeLists.txt:**
Added a library for the Nnapi delegate
- Explicit linking of torch_python is necessary for the Nnapi delegate's use of pybind

**test_backends.py:**
Added a test for lowering to Nnapi
- Based off https://github.com/pytorch/pytorch/blob/master/test/test_nnapi.py
- Only differences are the loading of the nnapi backend library and the need to change dtype from float64 to float32

### Test Plan:
Running `python test/test_jit.py TestBackendsWithCompiler -v` succeeds. Also saved and examined the model file locally.

Reviewers:

Subscribers:

Tasks:
T91991864
Tags:

Differential Revision: [D29687143](https://our.internmc.facebook.com/intern/diff/D29687143)

[ghstack-poisoned]

Amyh11325 commented:
@Amyh11325 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29687143).

raziel reviewed (commented):
Thanks

malfet commented on test/cpp/jit/CMakeLists.txt:
```diff
@@ -17,6 +17,16 @@ add_library(backend_with_compiler SHARED
 target_link_libraries(backend_with_compiler torch)
 
+# Don't link nnapi_backend library on MacOS
+if(NOT ${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
```
This change should have been guarded with `AND BUILD_PYTHON`)

malfet replied:
Also, this is a bit weird place to keep build definitions for library that depends on libtorch_python

malfet commented:
Reverting, as it broke non-python (namely libtorch_cpp) builds, and would probably affect number of Mobile builds as well.

All build rules that depends on Python developer headers should be guarded with `if(BUILD_PYTHON)`
